nifi-nar-bundles/nifi-questdb-bundle/nifi-questdb/src/main/java/org/apache/nifi/questdb/embedded/EmbeddedClient.java

```diff
@@ -0,0 +1,151 @@
+
+        try (final SqlCompiler compiler = getCompiler()) {
+            final CompiledQuery compile = compiler.compile(query, getSqlExecutionContext());
+            compile.execute(new SCSequence(new TimeoutBlockingWaitStrategy(5, TimeUnit.SECONDS)));
+        } catch (final SqlException | CairoError e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+    @Override
+    public void insert(
+        final String tableName,
+        final InsertRowDataSource rowDataSource
+    ) throws DatabaseException {
+        checkConnectionState();
+
+        if (!rowDataSource.hasNextToInsert()) {
+            LOGGER.debug("No rows to insert into {}", tableName);
+            return;
+        }
+
+        final TableToken tableToken = engine.get().getTableTokenIfExists(tableName);
+
+        if (tableToken == null) {
+            throw new DatabaseException(String.format("Table Token for table [%s] not found", tableName));
+        }
+
+        try (
+            final TableWriter tableWriter = engine.get().getWriter(tableToken, "adding rows")
+        ) {
+            final TableWriterBasedInsertRowContext context = new TableWriterBasedInsertRowContext(tableWriter);
+
+            while (rowDataSource.hasNextToInsert()) {
+                context.addRow(rowDataSource);
+            }
+
+            LOGGER.debug("Committing {} rows", tableWriter.getRowCount());
+            tableWriter.commit();
+        } catch (final Exception | CairoError e) {
+            // CairoError might be thrown in extreme cases, for example when no space left on the disk
+            throw new DatabaseException(e);
+        } finally {
+            engine.get().releaseInactive();
+        }
+    }
+
+    @Override
+    public <T> T query(final String query, final QueryResultProcessor<T> rowProcessor) throws DatabaseException {
+        checkConnectionState();
+
+        final CompiledQuery compiledQuery;
+
+        try (final SqlCompiler compiler = getCompiler()) {
+            compiledQuery = compiler.compile(query, getSqlExecutionContext());
+        } catch (final SqlException | CairoError e) {
+            throw new DatabaseException(e);
+        }
+
+        try (
+            final RecordCursorFactory factory = compiledQuery.getRecordCursorFactory();
+            final RecordCursor cursor = factory.getCursor(getSqlExecutionContext());
+        ) {
+            final CursorBasedQueryRowContext rowContext = new CursorBasedQueryRowContext(cursor);
+
+            while ((rowContext.hasNext())) {
+                rowContext.moveToNext();
+                rowProcessor.processRow(rowContext);
+            }
+
+            return rowProcessor.getResult();
+        } catch (final Exception e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+    @Override
+    public void disconnect() throws DatabaseException {
+        checkConnectionState();
+        disconnected.set(true);
+        LOGGER.info("Client disconnects");
```

Comment:
In that case, recommend the following wording adjustment:
```suggestion
        LOGGER.info("Client disconnected");
```